tools/functional-tester/etcd-tester/checks.go

```diff
@@ -65,6 +90,85 @@ func (hc *hashChecker) Check() (err error) {
 	return nil
 }
 
+func (lc *leaseChecker) Check() error {
+	plog.Info("lease stresser invariant check...")
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for _, ls := range lc.leaseStressers {
+		err := lc.invariantCheck(ctx, ls)
+		if err != nil {
+			return fmt.Errorf("lease stresser encounters error: %v", err)
+		}
+	}
+	return nil
+}
+
+func (lc *leaseChecker) invariantCheck(ctx context.Context, lStresser Stresser) error {
+	// invariantCheck waits until the current run of lease stresser finishes before starting to check invariants
+	ls := lStresser.(*leaseStresser)
+	// wait until lease stresser has finished run() before starting the invariant check
+	ls.runWg.Wait()
+	for leaseID := range ls.revokedLeases {
+		keysExpired, err := ls.hasKeysAttachedToLeaseExpired(ctx, leaseID)
+		if err != nil {
+			plog.Errorf("hasKeysAttachedToLeaseExpired error %v", err)
+			return err
+		}
+
+		leaseExpired, err := ls.hasLeaseExpired(ctx, leaseID)
+		if err != nil {
+			plog.Errorf("hasLeaseExpired %v", err)
+			return err
+		}
+
+		if !leaseExpired && keysExpired {
+			return fmt.Errorf("revoked lease %v is not expired but attached keys are expired", leaseID)
+		}
+
+		if leaseExpired && !keysExpired {
+			return fmt.Errorf("revoked lease %v is expired but attached keys are not expired", leaseID)
+		}
+
+		if !leaseExpired && !keysExpired {
+			return fmt.Errorf("revoked lease %v is not expired and attached keys are not expired", leaseID)
+		}
+	}
+
+	ls.mu.Lock()
+	ls.revokedLeases = make(map[int64]struct{})
+	ls.mu.Unlock()
+
+	if len(ls.aliveLeases) != 0 {
+		// we wait for 2 * TTL to ensure that keepLeaseAlive() does indeed keep live leases alive
+		time.Sleep(TTL * 2 * time.Second)
+	}
+
+	for leaseID := range ls.aliveLeases {
```

Comment:
this is essentially the same loop as above; make the loop a function parameterized by the lease list and a boolean if it's expected to be expired?